lib/mongo/monitor.ex

```diff
@@ -1,49 +1,136 @@
+  # see https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#minheartbeatfrequencyms
+  @min_heartbeat_frequency_ms 500
 
-  def start_link() do
-    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  def start_link(args, gen_server_opts \\ []) do
+    GenServer.start_link(__MODULE__, args, gen_server_opts)
   end
 
-  def add_conn(pid, pool, version) do
-    GenServer.call(__MODULE__, {:add_conn, pid, pool, version})
+  def stop(pid) do
+    GenServer.stop(pid)
   end
 
-  def wire_version(conn) do
-    [{^conn, version}] = :ets.lookup(@ets, conn)
-    version
+  def force_check(pid) do
+    GenServer.call(pid, :check, :infinity)
   end
 
-  def init([]) do
-    :ets.new(@ets, [:named_table, read_concurrency: true])
-    {:ok, %{monitors: %{}}}
+  ## GenServer callbacks
+
+  @doc false
+  def init([server_description, topology_pid, heartbeat_frequency_ms,
+            connection_opts]) do
+    opts = # monitors don't authenticate and use the "admin" database
+      connection_opts
+      |> Keyword.put(:database, "admin")
+      |> Keyword.put(:skip_auth, true)
+    {:ok, pid} = DBConnection.start_link(Mongo.Protocol, opts)
+    :ok = GenServer.cast(self, :check)
+    {:ok, %{connection_pid: pid, topology_pid: topology_pid,
+            server_description: server_description,
+            heartbeat_frequency_ms: heartbeat_frequency_ms}}
+  end
+
+  @doc false
+  def handle_cast(:check, state) do
+    check(state)
+  end
+
+  @doc false
+  def handle_call(:check, _from, state) do
+    check(state)
+  end
+
+  @doc false
+  def handle_info(:timeout, state) do
+    check(state)
   end
 
-  def handle_call({:add_conn, pid, pool, version}, _from, state) do
-    :ets.insert(@ets, {pid, version})
-    ref = Process.monitor(pid)
-    state = put_in(state.monitors[ref], pid)
+  ## Private functions
 
-    state =
-      if pool && :ets.lookup(@ets, pool) == [] do
-        :ets.insert(@ets, {pool, version})
-        ref = Process.monitor(pool)
-        put_in(state.monitors[ref], pool)
-      else
-        state
-      end
+  defp check(state) do
+    diff = :os.system_time(:milli_seconds) -
+           state.server_description.last_update_time
+    if diff < @min_heartbeat_frequency_ms do
+      {:noreply, state, diff}
+    else
+      server_description =
+        is_master(state.connection_pid, state.server_description)
 
-    {:reply, :ok, state}
+      :ok = GenServer.call(state.topology_pid,
+                           {:server_description, server_description}, 30_000)
+      {:noreply, %{state | server_description: server_description},
+       state.heartbeat_frequency_ms}
+    end
   end
 
-  def handle_info({:DOWN, ref, :process, _pid, _reason}, state) do
-    pid = state.monitors[ref]
-    :ets.delete(@ets, pid)
-    monitors = Map.delete(state.monitors, ref)
+  defp call_is_master(conn_pid) do
+    start_time = System.system_time
```

Comment:
`System.system_time` should not be used since it can change and warp. Use `System.monotonic_time` or `:timer.tc`.